Readme.md
PBS-GESTOR
========================================================================

PBS Gestor is a daemon that reads the PBS accounting log messages and
sends them to PostgreSQL database. Once started, it will continue running
continuously in background, exiting only if killed/terminated by user
or the connection to database is lost.

To check that it is running, type 'top -c' or 'ps -ef' and look for 'pbs_gestor'.

To check results, look at the contents of PostgreSQL database, schema and tables:
'pbsjoblogsdb', 'schema' and 'pbsjob', 'pbsjobarr', correspondingly, by default,
but these names are settings to be set by user in the configuration file.
The daemon runs the infinite loop within pbsloghandler.py::readline()

------------------------------------------------------------------------
INTRODUCTION TO PEX
------------------------------------------------------------------------

The PBS_Gestor project provides a *.pex file - which is a Python executable,
 self-sufficient and requiring no installation of Python dependencies.
See: https://pex.readthedocs.io/en/stable/index.html 

PEX has been created successfully on Debian and RedHat-based systems
 and should work on other Linux systems as well.
The way to run the *.pex is described in 'Run' section below.

The following are instructions on building a *.pex file.
If you just want to run a *.pex file, you can safely skip to the 'Install' section.

To build a *.pex file, you need to install the python package 'pex', first.

The PBS_Gestor project is packaged in a standard manner 
using Python's setupscript setup.py and setup.cfg .
This makes it easy for 'pex' to package the Python code and all the required dependencies.

The instructions in the Makefile create a *.pex file packaged for multiple versions of Python.

To build for a specific Python version, please add to Makefile a specific instruction as follows,
- for example for Python 2.7:

ccbuild:
- 	rm -rf pbs_gestor.egg-info
- 	python2 -m pip wheel -w . . --isolated
- 	python -m pex . -f ./ -v -c gestor.py -o pbs_gestor_py2.7.8.pex --disable-cache --no-compile --platform="linux_x86_64-cp-27-cp27mu"
- 	rm -rf pbs_gestor.egg-info

To make, run:
- make ccbuild

To make pbs_gestor be seen as process name on top, edit *.pex file in binary mode, for example:
- vim -b ./pbs_gestor_py2.7.8.pex

Change the top line to the correct, full Python path (which will be different for each system):
- #!/opt/miniconda3/envs/bom-schema-py2/bin/python

OR
- #!/usr/bin/python

Alternatively, instead of editing binary *.pex file, run top as:
- top -c

------------------------------------------------------------------------
ABOUT SPHINX DOCUMENTATION
------------------------------------------------------------------------

* In addition to other setup requirements listed in setupscript setup.py,
   the Sphinx documentation building process requires the following packages:
  - Sphinx
  - rst2pdf
  
  How to make Sphinx documentation:
  - cd docs
  - make docs
* If the build process is unsuccessful and gives error about 'sphinx-build' being not found,
   then installation of Python package 'Sphinx' is required, as mentioned above.
* If the build process is successful, the pdf will be found in docs/build/ directory.
* Error message:
   - mkdir ../tmp
   - mkdir: cannot create directory `../tmp': File exists
   - make: *** [docs] Error 1

  during process of building documentation indicates
  that a previous failed attempt of making documentation has left files in incorrect location.
  
  The solution is:
   - cd docs
   - mv ../tmp/*.py ../
   - rmdir tmp
   - make docs
* Error message:
   - WARNING: autodoc: failed to import module u'gestor'; the following exception was raised:
   - Traceback (most recent call last):
   - TypeError: __init__() takes exactly 2 arguments (3 given)

  during generation of documentation indicates that the code was written for psycopg2 2.8 or higher;
  and running the code with psycopg2 version 2.7 or below causes error,
  because function "Identifier" takes only one argument in the older version of psycopg2.
  
  Solution:
   - Besides updating psycopg2, we can instead modify utils.py to specify schema explicitly,
     such as,
     - '{schema}.{table}'.format(schema=Identifier('schema'), table=Identifier('table'))

     instead of
     - '{table}'.format(table=Identifier('schema', 'table'))

------------------------------------------------------------------------
INSTALL PBS GESTOR
------------------------------------------------------------------------

PRE-REQUISITES:
* Python, version 2.7 or 3.6: https://www.python.org/
* PBS server accounting logs access
* PostgreSQL, version 9, such as 9.2.24 or 9.6.11: https://www.postgresql.org/

PRE-SETUP:

* PBS:
  - Start the PBS Server.
* PostgreSQL:
  - Set up host-based authentication (HBA) configuration as needed
    usually in /var/lib/pgsql/data/pg_hba.conf
  - Start PostgreSQL server.
  - Add PostgreSQL authentication data to PBS Gestor configuration file pbs_gestor_config.json ,
    such as: hostname, port, username, password, database name;
  - When you run PBS Gestor, it determines where to look for the file
    (default system location such as ~/.config/pbs_gestor/pbs_gestor_config.json ,
    or PBS_GESTOR_CONF if such environmental variable is set),
    prints out the location of configuration file to stdout,
    and if such file does not exist (for example, on the first run of PBS Gestor),
    then copies the default template to that location.
  - Besides username, password and database name, the configuration file also contains
    extra/back-up values 'supdatabase', 'supusername' and 'suppassword',
    which are not used by the program unless necessary, see examples below.
  - If the database doesn't exist, it will be created - user needs to have CREATEDB permissions,
    and it will be done by briefly connecting to 'supdatabase' ("postgres" by default) first .
  - If tablefunc extension isn't installed in the database, it will be created - by
    briefly connecting as super-user (see 'supusername' being "postgres" by default) first.
  - If tablefunc extension is installed in the database, but in a schema inaccessible to user,
    user will be granted access to the schema by briefly connecting as super-user first.
  - Tablefunc extension is required for Crosstab function to pivot a table,
    see details on PbsJobArr table and PbsRawFlatView view below.

------------------------------------------------------------------------
SETUP
------------------------------------------------------------------------

* Copy or move the `gestor` directory to any location accessible/preferable for you.

* If you haven't run the program before, run the program once, so that it writes out
  the default settings to a configuration file, and prints out location of the configuration file,
  with either of the following commands:
  - /path/to/pbs_gestor.pex
  - python /path/to/pbs_gestor.pex
  - /path/to/pbs_gestor.pex -c
  - python /path/to/pbs_gestor.pex -c

    On the first run - defined as when program doesn't find the config file at expected location,
  and has to create the configuration file with default settings - or if run with '-c' flag,
  the program doesn't establish a database connection,
  but just prints out location of the configuration file and exits, 
  so that user can put in the specific settings.

* The following components needs to be updated before starting the PBS Gestor daemon:
    - pbs_gestor_config.json - system related config needed for PBS
    Gestor plugin. Includes database authentication, logging configuration.
    Add PostgreSQL authentication data here, such as:
    hostname, port, username, password, database name

    Note - If you move pbs_gestor_config.json to a different location,
    please create or update the environment variable PBS_GESTOR_CONF 
    with the location of pbs_gestor_config.json .

------------------------------------------------------------------------
RUN
------------------------------------------------------------------------

* Start the Daemon script with one of the following commands:
  - /path/to/pbs_gestor.pex

  which should work as long as Python interpreter can be found at /usr/bin/python
  - python /path/to/pbs_gestor.pex
  - python3 /path/to/pbs_gestor.pex
  - python3.x /path/to/pbs_gestor.pex

  If you have source code of the program, it can also be run similarly with:
  - python /path/to/gestor.py

  but that requires installation of dependencies such as sqlalchemy - full list is inside setup.py

* Once started, the program can be seen in list of processes, such as in 'top -c' or 'ps -ef'.
  
  If started directly as:
  - /path/to/pbs_gestor.pex

  then the process name will be the same as name of the file, for example, 'pbs_gestor.7.8.',
  and it can be terminated with, for example, 'pkill pbs_gestor'.
  
  If started indirectly, such as:
  - python /path/to/pbs_gestor.pex

  then the process name will be the same as name of the Python interpreter used to run it,
  for example, 'python3', but it can still be identified with, for example,
  'ps aux | grep pbs_gestor'.

* To see help about the program, and examples of usage, run it with ' --help', for example:
  - /path/to/pbs_gestor.pex --help

  For example, you can supply the dates between which to process the logs as follows:
  - /path/to/pbs_gestor.pex -f lastscan -t today
  - /path/to/pbs_gestor.pex -f 20011231 -t 20190726

  Keep in mind that once processing of logs between the specified dates is finished,
  program doesn't exit; it opens today's log file and starts scanning today's logs,
  continuing indefinitely to process jobs' records as they arrive.

* The program outputs to stdout some minimal amount of information,
  such as: the range of dates in which it is going to process logs,
  and which log it is processing. Far more verbose output can be found in
  log file, which is usually written to /tmp/pbs_gestor_log - the setting can be configured
  inside pbs_gestor_config.json , under logger / handlers / rotating_file_handler.

------------------------------------------------------------------------
DATABASE
------------------------------------------------------------------------

* Run some pbs jobs - If all the configurations are set properly then
  you should see some records in the PostgreSQL database. The jobs are recorded
  in the tables 'pbsjob' and 'pbsjobarr' under schema 'schema' ,
  inside the database 'pbsjoblogsdb'; all these names are settings 
  specified inside pbs_gestor_config.json .

  - PbsJob table contains jobs and their attributes:
    name of the job, running time, vnode, user who started it, et cetera.
  - PbsJobArr contains job resources: assigned to jobs, or used by jobs.
      One column contains unique job identifier for relationship to PbsJob table,
      another column contains name of resource, and the last column contains value of resource.
  - Supplementary table PBSLog contains log handler records of which logs were processed when.
  
  Upon connecting to the PostgreSQL database
  - for example, with the command
   - psql -h hostname -p port -U username  -d databasename
   - which with default database connection settings would be
   - psql -h localhost -p 5432 -U postgres  -d pbsjoblogsdb

  you would be able to query the contents of the tables inside the database, such as:
  - SELECT * FROM schema.pbsjob;

  OR
  - SELECT * FROM schema.pbsjob WHERE ji_user='username';

  where username is name of a user whose job history you want to look up
  OR
  - SELECT count(*) FROM schema.pbsjob WHERE ji_user='username';

  if you want to count how many jobs the user has run in the past
  OR
  - any other SQL query
  
  Additionally, there are also two Views created, 'pbsrawflatview' and 'pbsrawflatviewjoin',
  inside the same schema as tables.
  - PbsFlatView contains a subset of information from PbsJobArr,
    pivoting the table with PostgreSQL's Crosstab function from Tablefunc extension
    so that requested and used resources would be organised into the following columns:
    'mem', 'ncpus', 'walltime', 'cput', 'nodect', 'cpupercent', et cetera
    instead of being contained in a Entity–attribute–value model as in PbsJobArr table.

    See: https://www.postgresql.org/docs/9.2/tablefunc.html#AEN152349

    Resources requested by job are prefixed with l,
    to be differentiated from resources used by job.

    It is also joined with PbsJob table so that Job ID would be available as one of the columns.
    See INNER JOIN on the following documentation page:

    https://www.postgresql.org/docs/9.2/queries-table-expressions.html#QUERIES-JOIN

    The values for the resources, such as ncpus or walltime, are cast to appropriate data types,
    so that arithmetic operations can be performed on them.

  These views can be queried similarly to tables, for example:
  - SELECT * FROM schema.pbsflatviewreqjoin WHERE l_nodect > 1;

  displays list of all jobs which requested/used more than one node.

------------------------------------------------------------------------
TROUBLESHOOT
------------------------------------------------------------------------

* For troubleshooting, try to connect to the database with the same settings
  (hostname, port, username, databasename) as are inside pbs_gestor_config.json file.
  - For ease of troubleshooting, application prints out the location from which it is reading
    the configuration file pbs_gestor_config.json . As noted earlier, if you move the file
    to a different location, please create or update the environment variable PBS_GESTOR_CONF 
    with the location of pbs_gestor_config.json .
  
  Check whether schema 'schema' exists by querying list of schemas with '\dn'
  command inside psql client.
  
  Check whether tables exist with '\dt schema.*' command.
  
  Check whether tablefunc extension is installed with the following command:
  - \dx+ tablefunc

  If it says 'Did not find any extension named "tablefunc"',
  then Tablefunc extension is not installed and Crosstab function is not available.
  If the Tablefunc extension is installed successfully,
  then there will be list of objects in extension "tablefunc",
  such as connectby, crosstab and normal_rand.
  OR
  - SELECT count(*) FROM information_schema.routines WHERE routine_name LIKE 'crosstab%'

  If it is 0, then Tablefunc extension is not installed and Crosstab function is not available.
  If it is above 0 - for example, 6 - then Crosstab function is available, which is good.

  If Tablefunc extension is installed, then you may also wish to check where it is installed:

release_notes.txt
========================================================================
PBS-GESTOR
========================================================================

PBS Gestor is a daemon that reads the PBS accounting log messages and
sends them to the PostgreSQL database in the desired format.

------------------------------------------------------------------------
Package v7.9
------------------------------------------------------------------------
* Fixed logging bug.
* More detailed exception logging.
* More detailed release notes.
* Closing all connections in orm lib.

------------------------------------------------------------------------
Package v7.8
------------------------------------------------------------------------
* Compatibility with psycopg2 earlier than 2.8.
* If extension 'tablefunc' is installed in a different schema, use it, and give user access to the schema as required.

------------------------------------------------------------------------
Package v7.7
------------------------------------------------------------------------
* Extended documentation.
* Configurable names of schema and tables.

------------------------------------------------------------------------
Package v7.6
------------------------------------------------------------------------
* Combined two views into one, and cast columns from text to proper data type.
* Added 'Requested resources' to view.
* Fixed issue with links inside pdf Documentation.
* Changed treatment of config file so that database connection would not happen on the first run.

------------------------------------------------------------------------
Package v7.5
------------------------------------------------------------------------
* Added usage, examples and troubleshooting to install.txt and documentation.
* Configurable super-user name and password, and name of super-database.
* Bandit finds no issues, added to Makefile's 'lint' section so that it can be run with 'make lint'

------------------------------------------------------------------------
Package v7.4
------------------------------------------------------------------------
* Tested to work with non-superuser PostgreSQL user, as specified in the config file.
* For creation of extension (if needed), switches to "postgres" superuser with password "postgres".
* For creation of database (if needed), switches to "postgres" database.
* Tries both 'crosstab' and 'SCHEMA.crosstab' functions - in case the extension is installed inside SCHEMA, or outside it.

------------------------------------------------------------------------
Package v7.3
------------------------------------------------------------------------
* Fix the problem with crosstab function in tablefunc extension

------------------------------------------------------------------------
Package v7.2
------------------------------------------------------------------------
* Ability to process logs over a certain date range.
* Log table for recording previous runs
* Pylint Score - 10/10
* Documentation.

------------------------------------------------------------------------
Package v7.1
------------------------------------------------------------------------
* Integration of Producer and Consumer into one daemon.
* Flat views of tables with jobs and resources, with attributes such as walltime and CPU usage.
* Creates database, schema, tables and views if they do not exist yet.
* Python 2.7 - Python 3.7 compatibility.

NOTES:
* To build pex file, do (from 'gestor' directory): 'make buildallwheels';
  requires all versions of Python (2.7, 3.4, 3.5, 3.6, 3.7) installed and available in the $PATH,
  with pip, wheel and setuptools installed for each of the Python versions.
* To build documentation in PDF format, do (from 'gestor/docs' directory): 'make docs';
  requires Python packages 'Sphinx' and 'rst2pdf' installed, see https://www.sphinx-doc.org/ . 
  The resulting pdf file will be in 'gestor/docs/build/pdf/' directory.


pbs_gestor/utils.py
# -*- coding: utf-8 -*-

# Copyright (C) 1994-2018 Altair Engineering, Inc.
# For more information, contact Altair at www.altair.com.
#
# Commercial License Information:
#
# For a copy of the commercial license terms and conditions,
# go to: (http://www.pbspro.com/UserArea/agreement.html)
# or contact the Altair Legal Department.
#
# Use of Altair’s trademarks, including but not limited to "PBS™",
# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject
# to Altair's trademark licensing policies.
r"""
Hold structure of database, constants and utility functions to be used across PBS Gestor modules.

Include schema, tables and views.

Classes corresponding to tables:
    * PbsJob: This table contains jobs and their attributes:
      name of the job, running time, vnode, user who started it...
      - The table contains 'CONSTRAINT unq_pbsjob UNIQUE (ji_jobid)'
      so that duplicate records are not created for the same job.

    * PbsJobArr: This table contains job resources:
      assigned to jobs (l_* resource names), or used by jobs (u_* resource names).
      One column contains unique job identifier for relationship to PbsJob table,
      another column contains name of resource, and the last column contains value of resource;
      names and values of resources are represented as text.

    * PBSLog: This supplementary table contains log handler records
      of which logs were processed when.

    Upon connecting to the PostgreSQL database

    - for example, with the command
      psql -h hostname -p port -U username  -d databasename
      with default database connection settings it would be
      psql -h localhost -p 5432 -U postgres  -d pbsjoblogsdb
      you would be able to query the contents of the tables inside the database, such as:

    - SELECT * FROM schema.pbsjob;

    OR

    - SELECT * FROM schema.pbsjob WHERE ji_user='username';
      where username is name of a user whose job history you want to look up

    OR

    - SELECT count(*) FROM schema.pbsjob WHERE ji_user='username';
      if you want to count how many jobs the user has run in the past

    OR

    - any other SQL query

    Additionally, there is also a View created, with name specified in the configuration file,
    inside the same schema as tables.

    - PbsFlatView contains a subset of information from PbsJobArr,
      pivoting the table with PostgreSQL's Crosstab function from Tablefunc extension
      so that requested and used resources would be organised into the following columns:
      'mem', 'ncpus', 'walltime', 'cput', 'nodect', 'cpupercent', et cetera
      instead of being contained in a Entity–attribute–value model as in PbsJobArr table.
      See: https://www.postgresql.org/docs/9.2/tablefunc.html#AEN152349
      Resources requested by job are prefixed with l,
      to be differentiated from resources used by job.
      It is also joined with PbsJob table so that Job ID would be available as one of the columns.
      See INNER JOIN on the following documentation page:
      https://www.postgresql.org/docs/9.2/queries-table-expressions.html#QUERIES-JOIN
      The values for the resources, such as ncpus or walltime, are cast to appropriate data types,
      so that arithmetic operations can be performed on them.

    This view can be queried similarly to tables, for example:

    - SELECT * FROM schema.pbsflatviewreqjoin WHERE l_nodect > 1;
      displays list of all jobs which requested/used more than one node.

* For troubleshooting, try to connect to the database with the same settings
    (hostname, port, username, databasename) as are inside pbs_gestor_config.json file.

    - For ease of troubleshooting, application prints out the location from which it is reading
      the configuration file pbs_gestor_config.json . As noted earlier, if you move the file
      to a different location, please create or update the environment variable PBS_GESTOR_CONF
      with the location of pbs_gestor_config.json .

    Check whether schema 'schema' exists by querying list of schemas with '\\dn'
    command inside psql client.

    Check whether tables exist with '\\dt schema.*' command.

    Check whether tablefunc extension is installed with the following command:

    - \\dx+ tablefunc

      If it says 'Did not find any extension named "tablefunc"',
      then Tablefunc extension is not installed and Crosstab function is not available.
      If the Tablefunc extension is installed successfully,
      then there will be list of objects in extension "tablefunc",
      such as connectby, crosstab and normal_rand.

    OR

    - SELECT count(*) FROM information_schema.routines WHERE routine_name LIKE 'crosstab%';

      If it is 0, then Tablefunc extension is not installed and Crosstab function is not available.
      If it is above 0 - for example, 6 - then Crosstab function is available, which is good.

    If Tablefunc extension is installed, then you may also wish to check where it is installed:

    - SELECT routine_schema FROM information_schema.routines WHERE routine_name LIKE 'crosstab%';

    OR

    - SELECT specific_schema FROM information_schema.routines WHERE routine_name LIKE 'crosstab%';

      (these two similar commands usually return the same result)

    to see which schema contains the installed Crosstab function.

    - Application expects this result to be either 'schema' or, failing that,
      something like 'public' so that the function is inside PostgreSQL's search_path.
      See: https://www.postgresql.org/docs/9.2/runtime-config-client.html#GUC-SEARCH-PATH

Attributes:
    * CONFIGS_DIR_PATH: The path where all the configuration files
    required by PBS Gestor reside. The current codebase works with
    having the configs directory and having the json files like
    pbs_gestor_config.json
    * GESTOR_CONFIG: Dictionary of pbs gestor related config
    * DEFAULT_PBS_GESTOR_CONF: The default path of the the PBS Gestor
    configuration file - ~/.config/pbs_gestor/psb_gestor_config.json
    * LOGGING_CONFIG : The logger configuration for PBS Gestor
"""

from __future__ import print_function
import os
import json
from datetime import datetime
# try: Dependency problems when trying to package into multi-platform PEX
#     import importlib.resources as pkg_resources
# except ImportError:
#     # Try backported to PY<37 `importlib_resources`.
#     import importlib_resources as pkg_resources
import pkg_resources
import appdirs
# from .model.exceptions import ConfigurationError
# from .model import TABLES, SQLVIEWS
from sqlalchemy import Column, Text, Integer, DateTime, UniqueConstraint, ForeignKey
from sqlalchemy import BIGINT, ARRAY, Date
from sqlalchemy.orm import relationship
from psycopg2.sql import SQL, Identifier
from pbs_gestor.model.orm_lib import BASE
from pbs_gestor.model.orm_lib import Helpers


def get_config():
    """Find and read configuration file and return its contents."""
    try:
        configs_dir_path = appdirs.user_config_dir('pbs_gestor')
        default_pbs_gestor_conf = os.path.join(configs_dir_path, 'pbs_gestor_config.json')
        pbs_gestor_conf = os.getenv('PBS_GESTOR_CONF', default_pbs_gestor_conf)
        if not os.path.isfile(pbs_gestor_conf):
            first_run = True
            create_user_config(pbs_gestor_conf)
        else:
            first_run = False
        with open(pbs_gestor_conf) as pbs_gestor_config:
            print("Loading configuration from file %s." % pbs_gestor_conf)
            data = json.load(pbs_gestor_config)
            database = data["database"]
            if "supdatabase" not in database:
                database["supdatabase"] = "postgres"
            if "supusername" not in database:
                database["supusername"] = "postgres"
            if "suppassword" not in database:
                database["suppassword"] = "postgres"
            return data, first_run
    except Exception:
        raise


def create_user_config(config_file):
    """Create the user's config file."""
    src = pkg_resources.resource_stream(__name__, 'pbs_gestor_config.json')
    try:
        with open(config_file, 'wb+') as dest:
            dest.writelines(src)
    except IOError:
        os.makedirs(os.path.dirname(config_file))
        with open(config_file, 'wb+') as dest:
            dest.writelines(src)


GESTOR_CONFIG, GESTOR_FIRST_RUN = get_config()
LOGGING_CONFIG = GESTOR_CONFIG["logger"]
SCHEMA = GESTOR_CONFIG["schematables"]["schema"]
LOGTABLE = GESTOR_CONFIG["schematables"]["pbsgestorlogtable"]
JOBTABLE = GESTOR_CONFIG["schematables"]["pbsjobtable"]
JOBARRTABLE = GESTOR_CONFIG["schematables"]["pbsjobresourcestable"]
FLATVIEW = GESTOR_CONFIG["schematables"]["pbsjobresourceview"]

DATA_MAPPER = {
    "ji_cr_time": lambda x: datetime.isoformat(datetime.utcfromtimestamp(float(x))),
    "ji_quetime": lambda x: datetime.isoformat(datetime.utcfromtimestamp(float(x))),
    "ji_start_time": lambda x: datetime.isoformat(datetime.utcfromtimestamp(float(x))),
    "ji_end_time": lambda x: datetime.isoformat(datetime.utcfromtimestamp(float(x)))
}


class PbsJobArr(BASE):
    """Holds table structure: resources assigned to jobs, or used by jobs."""

    # pylint: disable=too-few-public-methods
    __tablename__ = JOBARRTABLE  # 'pbsjobarr'  # name
    __table_args__ = (UniqueConstraint('ji_pbsjobidx', 'ji_arrresource', 'ji_arrvalue'),
                      {"schema": SCHEMA, "extend_existing": True})  # schema
    attributes = ["ji_pbsjobidx", "ji_arrresource", "ji_arrvalue"]
    p_key = ["ji_pbsjobarridx"]

    # Table definition
    ji_pbsjobarridx = Column(Integer, primary_key=True, autoincrement=True)
    ji_pbsjobidx = Column(Integer, ForeignKey(Helpers.schema_ref(SCHEMA,  # schema  name
                                                                 '%s.ji_pbsjobidx' % JOBTABLE)),
                          nullable=False)
    ji_arrresource = Column(Text)
    ji_arrvalue = Column(Text)

    def __init__(self, **kwargs):  # , schema=SCHEMA, name='pbsjobarr', pbsjob='pbsjob'
        """Define columns."""
        self.ji_pbsjobidx = kwargs["ji_pbsjobidx"]
        self.ji_arrresource = kwargs["ji_arrresource"]
        self.ji_arrvalue = kwargs["ji_arrvalue"]

    def __str__(self):
        """String representation of the table."""
        _p_key = {
            'ji_pbsjobidx': self.ji_pbsjobidx,
            'arrresource': self.ji_arrresource,
            'arrvalue': self.ji_arrvalue
        }
        return "table :- {table}, p_key :- {attr}".format(table=self.__tablename__, attr=_p_key)


class PbsLog(BASE):
    """
    This class holds the logs of past runs of loghandler.

    These records are utilised to make sure
    that the same logs are not processed over and over again.
    """

    # pylint: disable=too-few-public-methods
    __tablename__ = LOGTABLE  # 'pbslog'  # name
    __table_args__ = {"schema": SCHEMA, "extend_existing": True}  # schema
    attributes = ["filename", "start", "end"]
    p_key = ["idx"]

# Table definition
    idx = Column(Integer, primary_key=True, autoincrement=True)
    start = Column(DateTime)
    end = Column(DateTime)
    filename = Column(Date, unique=True)

    def __init__(self, **kwargs):  # , schema=SCHEMA, name='pbslog'
        """Define columns."""
        self.start = kwargs["start"]
        self.end = kwargs["end"]
        self.filename = kwargs["filename"]

    def __str__(self):
        """String representation of the table."""
        _p_key = {
            'filename': self.filename
        }
        return "table :- {table}, p_key :- {attr}".format(table=self.__tablename__, attr=_p_key)


class PbsJob(BASE):
    """
    Hold the structure of table with jobs.

    Describe when a job was added into a queue,
    started, or finished, and other attributes (except resources).
    """

    # pylint: disable=too-many-instance-attributes,too-few-public-methods
    # Eighteen is reasonable in this case - considering number of columns in the table.
    __tablename__ = JOBTABLE  # 'pbsjob'  # name
    __table_args__ = {"schema": SCHEMA, "extend_existing": True}  # schema
    attributes = ["ji_jobid", "ji_jobname", "ji_user", "ji_group", "ji_project",
                  "ji_sv_name", "ji_queue", "ji_priority", "ji_cr_time", "ji_quetime",
                  "ji_runcount", "ji_eligible_time", "ji_start_time", "ji_end_time",
                  "ji_sessionid", "ji_exitstat", "ji_exechost", "ji_execvnode"]
    p_key = ["ji_pbsjobidx"]

    # Table definition
    ji_pbsjobidx = Column(Integer, primary_key=True, autoincrement=True)
    ji_jobid = Column(Text, nullable=False, unique=True)
    ji_jobname = Column(Text, nullable=False)
    ji_user = Column(Text, nullable=False)
    ji_group = Column(Text, nullable=False)
    ji_project = Column(Text, nullable=False)
    ji_sv_name = Column(Text, nullable=False)
    ji_queue = Column(Text, nullable=False)
    ji_priority = Column(Integer, nullable=False)
    ji_cr_time = Column(DateTime)
    ji_quetime = Column(DateTime)
    ji_runcount = Column(BIGINT)
    ji_eligible_time = Column(BIGINT)
    ji_start_time = Column(DateTime)
    ji_end_time = Column(DateTime)
    ji_sessionid = Column(Integer)
    ji_exitstat = Column(Integer)
    ji_exechost = Column(ARRAY(Text))
    ji_execvnode = Column(ARRAY(Text))
    pbsjobarr = relationship(PbsJobArr, backref=JOBTABLE)  # name

    def __init__(self, **kwargs):  # , schema=SCHEMA, name='pbsjob'
        """Define columns."""
        self.ji_jobid = kwargs["ji_jobid"]
        self.ji_jobname = kwargs["ji_jobname"]
        self.ji_user = kwargs["ji_user"]
        self.ji_group = kwargs["ji_group"]
        self.ji_project = kwargs["ji_project"]
        self.ji_sv_name = kwargs["ji_sv_name"]
        self.ji_queue = kwargs["ji_queue"]
        self.ji_priority = kwargs["ji_priority"]
        self.ji_cr_time = kwargs["ji_cr_time"]
        self.ji_quetime = kwargs["ji_quetime"]
        self.ji_runcount = kwargs["ji_runcount"]
        self.ji_eligible_time = kwargs["ji_eligible_time"]
        self.ji_start_time = kwargs["ji_start_time"]
        self.ji_end_time = kwargs["ji_end_time"]
        self.ji_sessionid = kwargs["ji_sessionid"]
        self.ji_exitstat = kwargs["ji_exitstat"]
        self.ji_exechost = kwargs["ji_exechost"]
        self.ji_execvnode = kwargs["ji_execvnode"]

    def __str__(self):
        """String representation of the table."""
        _p_key = {
            'jobid': self.ji_jobid
        }
        return "table :- {table}, p_key:- {attr}".format(table=self.__tablename__, attr=_p_key)


PBSRAWFLATVIEW = {"__schema__": SCHEMA,
                  "__tablename__": FLATVIEW}  # "pbsflatviewreqjoin"}
PBSRAWFLATVIEW[("__statement"
                "__")] = SQL("CREATE OR REPLACE VIEW {exschema}.{vname} AS "
                             "SELECT {exschema}.{pbsjob}.ji_jobid, "
                             "{tmp}.l_memory,{tmp}.l_ncpus::int,{tmp}.l_nodect::int,{tmp}.l_place,"
                             "{tmp}.l_vmem,{tmp}.l_walltime::interval,"
                             "{tmp}.cpu_percent::int,{tmp}.cpu_time::interval,{tmp}.memory,"
                             "{tmp}.ncpus::int,"
                             "{tmp}.vmem,{tmp}.walltime::interval FROM {exschema}.crosstab"
                             "('select ji_pbsjobidx,{res},ji_arrvalue from {exschema}.{pbsjobarr}"
                             " where {res}=''u_ncpus'' or {res}=''u_mem'' or"
                             " {res}=''u_vmem'' or {res}=''u_walltime'' or "
                             "{res}=''u_cput'' or {res}=''u_cpupercent'' or "
                             "{res}=''l_ncpus'' or {res}=''l_mem'' or {res}=''l_vmem'' or "  # vmem
                             "{res}=''l_walltime'' or "  # walltime
                             "{res}=''l_nodect'' or "
                             "{res}=''l_place'' order by 1,2;', "
                             "'VALUES (''l_mem''), (''l_ncpus''), (''l_nodect''), (''l_place''), "
                             "(''l_vmem''), (''l_walltime''), (''u_cpupercent''), (''u_cput''), "
                             "(''u_mem''), (''u_ncpus''), (''u_vmem''), (''u_walltime'')') "
                             "AS {tmp}(pbsjobidx int,"
                             "l_memory text,l_ncpus text,l_nodect text,l_place text,l_vmem text,"
                             "l_walltime text,"
                             "cpu_percent text,cpu_time text,memory text,ncpus text,"
                             "vmem text,walltime text) INNER JOIN {exschema}.{pbsjob} "
                             "ON {tmp}.pbsjobidx = {exschema}.{pbsjob}.ji_pbsjobidx"
                             ";").format(vname=Identifier(PBSRAWFLATVIEW["__tablename__"]),
                                         pbsjobarr=Identifier(JOBARRTABLE),
                                         pbsjob=Identifier(JOBTABLE),
                                         exschema=Identifier(SCHEMA),
                                         res=Identifier('ji_arrresource'),
                                         tmp=Identifier('pbsjobarrow'))

# Identifier(schema, table) is nice but works only since psycopg2 version 2.8

TABLES = {
    JOBTABLE: PbsJob,
    JOBARRTABLE: PbsJobArr,
    LOGTABLE: PbsLog
}


SQLVIEWS = {
    PBSRAWFLATVIEW["__tablename__"]: PBSRAWFLATVIEW
}
